fix: [audit] ZNS-8 Mintlist

Whytecrowe opened the pull request.

Remscar commented on contracts/oz/token/ERC721/ERC721Upgradeable.sol:
```diff
@@ -506,7 +506,8 @@ contract ERC721Upgradeable is
       ERC721Upgradeable.ownerOf(tokenId) == from,
       "ERC721: transfer of token that is not own"
     ); // internal owner
-    require(to != address(0), "ERC721: transfer to the zero address");
```
Would be good idea to rename this file with this change, so you have the old version still intact and can easily switch back to if needed.

JamesEarle replied:
Good idea

Remscar commented on contracts/registrar/ZNSHub.sol:
```diff
@@ -91,8 +91,6 @@ contract ZNSHub is
   // Beacon Proxy used by zNS Registrars
   address public beacon;
 
-  uint8 private test;
```
You can't remove variables from smart contracts once they're already there, unless you use special flags in the upgrade process

Remscar commented on scripts/rootMigration/actions/runSimulation.ts:
```diff
@@ -0,0 +1,173 @@
+  // Begin Migration simulation
+  // 2. Upgrade existing proxies to have needed functionality
+  // 2.a. Upgrade legacy Registrar
+  const upgradeRegistrarTx: MigrationRegistrar = await hre.upgrades.upgradeProxy(
```
You should be running the same code to migrate and simulate. Refactor the business logic from `runMigration` into a utility file, then import and run that code in `runSimulation`. If the code isn't the same, it's not a true simulation and you could have a bug in one, but not the other.

Remscar commented on scripts/rootMigration/actions/runSimulation.ts:
```diff
@@ -0,0 +1,173 @@
+  const upgradedBeacon = await upgradeBeaconTx.deployed();
+
+  // 3. Burn domains
+  await upgradedLegacyRegistrar.connect(signer).burnDomain(wilderDomainId);
```
You are going to want to `.wait(2)` on any and all transactions you do. When running in hardhat, you will want to only do `.wait()`

Remscar commented on scripts/rootMigration/actions/runSimulation.ts:
```diff
@@ -0,0 +1,173 @@
+
+  let tx_set;
+  // 6. Update the rootDomainId and parentRegistrar values on upgraded legacy
+  tx_set = await upgradedLegacyRegistrar.connect(signer).setRootDomainId(ethers.constants.HashZero);
```
Consider just making 'set root domain id' and 'set parent registrar' to be the same function

Remscar commented on scripts/rootMigration/actions/runSimulation.ts:
```diff
@@ -0,0 +1,173 @@
+  // 5. Mint 0://wilder in new root registrar
+  const tx = await beaconRegistrar.connect(signer).mintDomain(
+    ethers.constants.HashZero, // parent id
+    "0://wilder",
```
this should just be `wilder`

We don't include 0:// when minting domains, just the label

Remscar commented on scripts/rootMigration/postMigrate.ts:
```diff
@@ -0,0 +1,25 @@
+//    5a. Make these values point to the newly minted 0://wilder domain on beacon
+// 6. Upgrade both the legacy registrar and new beacon to have the old registrar code (PM)
+
+const migrate = async () => {
```
What's the point of this script? It looks like `runSimuation` already does this

JamesEarle replied:
meant to delete this, I had several before

Remscar commented on scripts/rootMigration/actions/runSimulation.ts:
```diff
@@ -0,0 +1,173 @@
+
+  console.log("new wilder domain: ", newWilderDomainId.toHexString());
+
+  let tx_set;
```
ultra nit pick: python naming convention on this variable

Remscar commented on scripts/rootMigration/actions/runSimulation.ts:
```diff
@@ -0,0 +1,173 @@
+    "ipfs.io/ipfs/Qm",
+    0,
+    false,
+    signerAddress,
```
Your `minter` and `sendToUser` are the same. Remove the send to user logic if you're not sending it to a user other than the minter

Remscar reviewed (changes requested):
Migration and simulation should be the exact same code, not copied code. It's dangerous to have them be different code since you might fix a bug in one but forget to fix a bug in the other. Also you will need to make sure you `.wait()` every transaction otherwise you'll get a failure. I would also recommend waiting for atleast 2-3 confirmations on mainnet.

If you want to make this safer, consider writing it so that it can be re-run without repeating steps in case of a failure mid-way through. For instance if we crash after upgrading the proxy beacon, let's just skip over that step. Might be much harder to code this, but could help prevent a sticky situation where you need to comment out code and re-run the script because of an transient error during the actual migration.

Remscar commented on contracts/oz/token/ERC721/ERC721PausableUpgradeable.sol:
```diff
@@ -2,7 +2,7 @@
 pragma solidity ^0.8.9;
 
-import "./ERC721Upgradeable.sol";
+import "./ERC721UpgradeableBurnable.sol";
```
Would be good to rename this file as well, like you did with UpgradeableBurnable.

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,143 @@
+    ]
+  );
+
+  const beaconRegistrar = await newRegistrarTx.deployed() as MigrationRegistrar;
```
Not contextual variable name. I'm assuming this is the new (0://) registrar?

JamesEarle replied:
yes, changed the name to `newRootRegistrar` to be clearer

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,143 @@
+    upgradedBeacon.address,
+    new MigrationRegistrar__factory(signer),
+    [
+      hre.ethers.constants.AddressZero, // Parent registrar
```
AddressZero technically is not the same as HashZero. This works, but it would be more correct to use HashZero.

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,143 @@
+    zNSHubAddress = zNSHub.address;
+    beaconAddress = beacon.address;
+
+    wilderDomainId = await helpers.mintSampleWilderDomain(signer, legacyRegistrar);
```
Why would you not use the `wilder` domain and instead use `wilder-sample` if we're doing a hardhat fork?

JamesEarle replied:
it was wilder before, I modified it to confirm the issue I DMd you about updating subdomain parents that wasn't an issue and just didn't change it back

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,143 @@
+
+  if (network === "hardhat") {
+    // Setup default contracts required for migration simulation
+    const [zNSHub, legacyRegistrar, beacon] = await helpers.deployContracts();
```
Why redeploy the contracts instead of use the existing contracts on mainnet? The hardhat network is a local fork, and can be configured to fork from mainnet, allowing for a true simulation.

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,143 @@
+  logger.log(`New Beacon Registrar Address: ${beaconRegistrar.address}`);
+
+  // Must do this to be able to mint below
+  await zNSHub.addRegistrar(rootDomainId, beaconRegistrar.address);
```
You are not waiting for this tx to complete

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  logger.log("4. Deploy new root registrar as proxy to existing subdomain registrar beacon");
+  const newRegistrarTx = await hre.upgrades.deployBeaconProxy(
```
You are not waiting for this tx to complete

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,143 @@
+  await zNSHub.addRegistrar(rootDomainId, beaconRegistrar.address);
+
+  logger.log("5. Mint wilder in new root registrar");
+  const tx = await beaconRegistrar.connect(signer).mintDomain(
```
You are not waiting for this tx to complete

JamesEarle replied:
It is being waited in the following function call to `getDomainId` but because that is no longer getting used more than once i've brought the wait directly in here

Remscar reviewed (changes requested):
You're still not waiting for all transactions to have block confirmations.
If you use libraries like `hre.upgrades` you need to figure out how to get the tx receipt and how to wait for block confirmations with those as well.

JamesEarle commented:
> You're still not waiting for all transactions to have block confirmations. If you use libraries like `hre.upgrades` you need to figure out how to get the tx receipt and how to wait for block confirmations with those as well.

For each use of a HardHat function for upgrading or deploying something we call `await newThing.deployed()` which internally calls to wait on the transaction then gives you the contract from there.
![image](https://user-images.githubusercontent.com/5976499/200968803-3dec96ec-abd4-48c8-a9fd-9358efc5e032.png)

Remscar commented on contracts/registrar/MigrationRegistrar.sol:
```diff
@@ -231,8 +229,12 @@ contract MigrationRegistrar is
 
-  function burnDomain(uint256 domainId) external onlyOwner {
-    _transfer(msg.sender, address(0), domainId);
+  function burnDomains(
```
I have a slight worry about a function like this. It shouldn't pose an issue but imagine if something goes wrong and you need to burn only a single domain. This function must burn two at a time; you can't burn just one domain.

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,184 @@
+    "ipfs://QmSQTLzzsPS67ay4SFKMv9Dq57iSQ7pLWQVUvS3XB5MowK/0",
+    0,
+    false,
+    newBeaconRegistrar.address
```
This is the wrong address. This should be the address of the wilder registrar (legacyRegistrarAddress)

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,184 @@
+    legacyRegistrarAddress,
+    new MigrationRegistrar__factory(signer)
+  );
+  const upgradedLegacyRegistrar = await upgradeRegistrarTx.deployed() as MigrationRegistrar;
```
You are still not properly waiting for these transactions to be confirmed. `.deployed()` waits only 1 block, you should wait `waitBlock` number of blocks by doing `upgradeRegistrarTx.deployTransaction.wait(waitBlocks)`

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,184 @@
+    new MigrationRegistrar__factory(signer)
+  );
+  // Beacon is the subdomain contract for wolf
+  const upgradedLegacyRegistrarBeacon = await upgradeBeaconTx.deployed() as MigrationRegistrar;
```
This is an invalid cast, this is not a `MigrationRegistrar` as you have casted it. It is a `ProxyBeacon` contract which you do not have any need to be holding a reference to.

JamesEarle replied:
Interesting, I expected that tx to return the type of contract we were upgrading, like `upgradeProxy` does but you're right. We don't use it so it didn't come up but good catch.

Remscar commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,184 @@
+    beaconAddress,
+    new MigrationRegistrar__factory(signer)
+  );
+  // Beacon is the subdomain contract for wolf
```
?

Whytecrowe commented on contracts/registrar/MigrationRegistrar.sol:
```diff
@@ -0,0 +1,810 @@
+   * Public View
+   */
+
+  // function ownerOf(uint256 tokenId)
```
Why is this commented out?

Whytecrowe commented on contracts/registrar/MigrationRegistrar.sol:
```diff
@@ -0,0 +1,810 @@
+    zNSHub.domainGroupUpdated(id, baseMetadataUri);
+  }
+
+  modifier onlyController() {
```
We seem to have 2 "functions" that check "if controller". This and `isController()` below. This seem to allow ZNSHub contract's controller to call, but not in `isController()`. It might be needed for the migration only and disallow the ZNSHub's controller to do it normally.
Just wanted to double-check to make sure both of them are needed.

Whytecrowe commented on contracts/registrar/MigrationRegistrar.sol:
```diff
@@ -0,0 +1,810 @@
+   */
+  function domainExists(uint256 id) public view override returns (bool) {
+    bool domainNftExists = _exists(id);
+    bool domainHasNotBeenBurned = ownerOf(id) != address(0);
```
This seems to use the function commented out above.

Whytecrowe commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,183 @@
+
+  logger.log("Comparing bytecode of Registrar and MigrationRegistrar for compatibility");
+  logger.log(`Registrar Bytecode: ${new Registrar__factory(signer).bytecode.length / 2}`);
+  logger.log(`MigrationRegistrar Bytecode: ${new MigrationRegistrar__factory(signer).bytecode.length / 2}`);
```
Why do we divide by 2 here?

Whytecrowe commented on scripts/rootMigration/actions/runMigration.ts:
```diff
@@ -0,0 +1,183 @@
+  logger.log(`Using network ${network} for migration`);
+
+  const rootDomainId = "0";
+  const waitBlocks = network === "hardhat" ? 0 : 3;
```
why we wait for 3 blocks?

Whytecrowe commented:
> If you want to make this safer, consider writing it so that it can be re-run without repeating steps in case of a failure mid-way through. For instance if we crash after upgrading the proxy beacon, let's just skip over that step. Might be much harder to code this, but could help prevent a sticky situation where you need to comment out code and re-run the script because of an transient error during the actual migration.

Second this! If possible, should look into it!

Whytecrowe commented on scripts/rootMigration/migrate.ts:
```diff
@@ -0,0 +1,21 @@
+  await hre.run("compile");
+
+  const [signer] = await hre.ethers.getSigners();
+  runMigration(signer, hre.network.name);
```
shouldn't this be awaited or returned?

deep-quality-dev commented on contracts/oz/token/ERC721/ERC721Upgradeable.sol:
```diff
@@ -506,7 +506,8 @@ contract ERC721Upgradeable is
     ); // internal owner
-    require(to != address(0), "ERC721: transfer to the zero address");
+    // To allow token burning we must be able to send to the zero address
+    // require(to != address(0), "ERC721: transfer to the zero address");
```
Burning token calls `_burn` function in `ERC721Upgradeable` contract, it will not call `_transfer` anymore.
We don't need to comment this `require` for burnning.

deep-quality-dev commented on contracts/registrar/MigrationRegistrar.sol:
```diff
@@ -0,0 +1,816 @@
+  ) external onlyOwner {
+    // ERC721 transfer will check owner internally
+    _transfer(_wilderOwner, address(0), _wilderDomainId);
+    _transfer(_rootOwner, address(0), _rootDomainId);
```
Burning should use `_burn` function instead of `_transfer`

deep-quality-dev commented on scripts/rootMigration/actions/runMigrationFromSafe.ts:
```diff
@@ -0,0 +1,426 @@
+    safeAddress,
+    signer,
+    rootRegistrar,
+    "mintDomain(uint256,string,address,string,uint256,bool,address)",
```
Before creating tx proposal to Gnosis Safe, `rootRegistrar` should transfer ownership to Gnosis Safe.
By default, [`hre.upgrades.deployBeaconProxy`](https://github.com/zer0-os/zNS/pull/60/files#diff-b43767899798227726329bb33b27e41ae4cd22276087ced4da127b52bf1f74dfR299) use deployer EOA as the owner of contract.

deep-quality-dev commented on scripts/rootMigration/actions/runMigrationFromSafe.ts:
```diff
@@ -0,0 +1,426 @@
+    data: data,
+    value: "0",
+    gasPrice: 1000000000, // 1 gwei
+    safeTxGas: 60000,
```
Is this gas fee enough to execute safe transaction?
I think it should be set by 0, so it can be reverted if internal tx is failed and use enough gas fees.
Please refer this link: https://github.com/safe-global/safe-contracts/blob/main/docs/safe_tx_gas.md

Whytecrowe committed:
implement initial clearMintlist function (not correct), combine with setting access type to locked for both Registrars

Whytecrowe committed:
rework logic to introduce a new struct with a mapping inside for mintlist

now instead of clearing a mapping, which we can't do, we change the idx of the internal mapping to a new one, creating a new mintlist respectively

Whytecrowe committed:
add/extend tests

Whytecrowe committed:
fix mintlist call in test

JamesEarle reviewed (approved):
nested mapping go brrrr :shipit:

Whytecrowe committed:
Merge branch 'development' into fix/audit-zns-8

zero-devops commented:
:tada: This PR is included in version 0.6.6 :tada:

The release is available on [GitHub release](https://github.com/zer0-os/zNS/releases/tag/v0.6.6-dev)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
